Src/java/cql-to-elm/src/test/resources/org/cqframework/cql/cql2elm/fhir/r401/TestInclude.cql

```diff
@@ -0,0 +1,170 @@
+library TestInclude
+
+using FHIR version '4.0.1'
+
+include FHIRHelpers version '4.0.1'
+
+codesystem "SNOMED": 'http://snomed.info/sct'
+codesystem "RXNORM": 'http://www.nlm.nih.gov/research/umls/rxnorm'
+
+code "aspirin 325 MG / oxycodone hydrochloride 4.84 MG Oral Tablet": '848768' from "RXNORM" display 'aspirin 325 MG / oxycodone hydrochloride 4.84 MG Oral Tablet'
+code "Acute hepatic failure": '197270009' from "SNOMED" display 'Acute hepatic failure'
+code "HIV Negative": '165815009' from "SNOMED" display 'HIV Negative'
+
+context Patient
+
+// Search Parameters
+
+define TestPractitionerSearch1:
+  [Practitioner: "?name" = 'ABC']
+
+define TestPractitionerSearch1A:
+  [Practitioner] P
+    where P."?name" = 'ABC'
+
+/*
+// [base]/Practitioner?name='ABC'
+
+define TestObservationEncounterSearch1:
+  [Observation: encounter = FHIRHelpers.reference('XYZ')]
+
+// Expected FHIR URL for the retrieve
+// [base]/Observation?subject=123&encounter=XYZ
+
+define TestMedicationSearch1:
+  [MedicationRequest: medication = FHIRHelpers.reference('XYZ')]
+
+// Expected FHIR URL for the retrieve
+// [base]/MedicationRequest?subject=123&medication=XYZ
+
+// TODO: Optimize to a search
+define TestMedicationSearch2:
+  [MedicationRequest] MR
+    with [Medication] M such that MR.medication = M.reference()
+
+// Chaining
+
+define TestMedicationRequestChaining1:
+  [MedicationRequest: medication.code ~ "aspirin 325 MG / oxycodone hydrochloride 4.84 MG Oral Tablet"]
+
+// Expected FHIR URL for the retrieve
+// [base]/MedicationRequest?subject=123&medication.code=http://www.nlm.nih.gov/research/umls/rxnorm|848768
+
+define TestEncounterChaining1:
+  [Encounter: diagnosis.code ~ "Acute hepatic failure"]
+
+// NOTE: This is the only reasonable translation here, but it is not right or useful or correct
+// Really it ought to be split into two retrieves, one for diagnoses with a condition and one for diagnoses with a procedure
+// [base]/Encounter?subject=123&diagnosis:Condition.code=http://snomed.info/sct|197270009&diagnosis:Procedure.code=http://snomed.info/sct|197270009
+
+// Reverse Chaining
+
+// Expected FHIR URL for the retrieve
+// [base]/Encounter?_has:Observation:context:code=http://snomed.info/sct|165815009
+
+// Includes
+*/
+
+// If a reference is resolved in the scope of a query, that resolve can be rewritten as an include in the retrieve for that source
+// MedicationRequest
+define TestMedicationRequest1:
+  [MedicationRequest] MR
+    where MR.medication.reference.resolve().as(Medication).code ~ "aspirin 325 MG / oxycodone hydrochloride 4.84 MG Oral Tablet"
```

Comment:
No, same answer here, the translator is supporting the planner with the references to the "resolve()" and "reference()" operators, but would stay out of the planning process because it doesn't have enough information to make an informed decision about the target environment.

Reply:
Documentation added to the TestIncludes library